README.md
# Save to the Wayback Machine
![Screenshot](assets/screenshots/readme-screenshot.png)

Browser extension for quickly saving web pages to the Internet Archive's [Wayback Machine](https://web.archive.org).

### Features
* Detects if a page can or can not be archived.
* See the last time a page was archived.
* Save pages, images and links via the right click menu.
* Open archived versions of pages, images and links via the right click menu.
* Keep track of the number of pages that you have archived.

## Download
[![Download from Firefox add-ons](assets/badges/firefox.png)](https://addons.mozilla.org/firefox/addon/save-to-the-wayback-machine/)
[![Download from the Google Chrome Web Store](assets/badges/chrome.png)](https://chrome.google.com/webstore/detail/save-to-the-wayback-machi/eebpioaailbjojmdbmlpomfgijnlcemk)
[![Download from Microsoft Edge Addons](assets/badges/ms-edge-chromium.png)](https://microsoftedge.microsoft.com/addons/detail/iooklpgaeiailiicahcjomphppicpgoo)

[Privacy Policy](PRIVACY.md) - [Browser Permissions](PRIVACY.md#browser-permissions) - [Installing in other browsers](INSTALL.md)

## Dependencies
- [Spacetime - JavaScript timezone library (version 6.16.2)](https://github.com/spencermountain/spacetime)

## Changelog

All notable changes to this project are documented in the [CHANGELOG](CHANGELOG.md).

## License

Code in this repository is licensed under the GNU General Public License Version 3. Details can be found in the [LICENSE](LICENSE) file. 

## Disclaimer

This is an unofficial extension and is not sponsored, owned, or endorsed by the [Internet Archive](https://archive.org).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [5.6.1](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/releases/tag/v5.6.1) - 2022-04-08
### Fixes:
* Fixed popup URL in `background.js`. ([`46af3c6`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/46af3c675cc67efa75455168ef216461502c6e7a))

## [5.6.0](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/releases/tag/v5.6.0) - 2022-04-08
### Additions:
* Added end-of-life message & page. ([#185](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/pull/185))

## [5.5.2](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/releases/tag/v5.5.2) - 2021-07-30
### Changes:
* Updated Spacetime timezone library to version [6.16.2](https://github.com/spencermountain/spacetime/releases/tag/6.16.2).

## [5.5.1](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/releases/tag/v5.5.1) - 2021-05-27
### Changes:
* Updated Spacetime timezone library to version [6.16.1](https://github.com/spencermountain/spacetime/releases/tag/6.16.1).

## [5.5.0](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/releases/tag/v5.5.0) - 2021-04-07
### Additions:
* Added keyboard accessible checkboxes and radio buttons. ([#179](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/pull/179))
* Added CSS `focus` rule to `select` drop-downs and buttons ([`644e118`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/644e1180785ec5f53a4bff51c85dbf55c5e2f2e7), [`862812f`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/862812f05bf60b276f7dd01483de12d02b0d1649))

### Changes:
* Updated Spacetime timezone library to version [6.15.0](https://github.com/spencermountain/spacetime/releases/tag/6.15.0).
* Buttons for archiving and viewing archived pages in `popup.html` are now disabled when not needed. ([`7855c3e`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/7855c3e388f573de63faeb0d0202dc41c7b011ff))
* Only lower opacity of `#overlay` background not its contents in `popup.css`. ([`6907465`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/6907465386a5de75afeddf33c3898fd531b1fd81))
* Minor changes to CSS rules in `popup.css`. ([`2f82a19`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/2f82a1904d697ff60891ec5c0deb5b501ec3bb8b))
* Removed unused and redundant CSS rules from `options.css` ([`ab064b1`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/ab064b1dafaf1ad1367b4e8024a55b51861b01e9), [`c8399a3`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/c8399a39c1736e2a531abede23d9cc41def677bc))

## [5.4.11](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/releases/tag/v5.4.11) - 2021-03-23
### Additions:
* Added function `hide()` to `ui.js`. ([`36f73fd`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/36f73fd794ea31d3698c398b396b9e4fed5441ea))
* Added english locale strings `PageArchived` and `PageNotArchived`. ([`b836a24`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/b836a24eda3385761d057e0378c050b6a27529c8), [`22c4684`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/22c46842a4b085d0f0e7911e7502f862cb3b67c1))
* Added english locale string `ApiPageNotArchived`. ([`a287044`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/a2870441c291fc0ec07db5cf4a5d9049e111bd19))

### Changes:
* Updated Spacetime timezone library to version [6.14.0](https://github.com/spencermountain/spacetime/releases/tag/6.14.0).
* Updated `popup.js` to set date & time in title attribute of HTML element `#since` when displaying time since in `popup.html`. ([`b21e27e`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/b21e27eab208f558d1c044b74a3c8629973b8d8a))
* Updated functions in `ui.js` to use 'typeof' when checking for undefined variables. ([`d96f183`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/d96f183bfaf99f7be6eed57e62ead60060ef39d0))
* Updated function `snapshotData()` to use a try...catch block in `popup.js` ([`03ef622`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/03ef622917fc1aa68a48c5662bc451f511392a80))
* Renamed array `hostNameBlacklist` to `hostnameDenylist` in `global.js` ([`9ef1603`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/9ef1603471be7e475fc419acf9dfa9a90a2347ca))
* Removed unused english locale string `StatisticsDisabled` ([`53f3e45`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/53f3e45c1f97b366b55a5f878f724999d09e13b3))

### Fixes:
* Fixed incorrect variable used in function `title()` in `ui.js`. ([`ca79dc3`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/ca79dc3acfa62758e142b1c85f55ebdb37265d8b))

## [5.4.10](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/releases/tag/v5.4.10) - 2021-02-16
### Changes:
* Updated Spacetime timezone library to version [6.12.5](https://github.com/spencermountain/spacetime/releases/tag/6.12.5).
* Updated privacy policy link in `options.html` to use [policy](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/blob/master/PRIVACY.md) hosted in the github repository. ([`2c08af`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/2c08af1c22953aaefba6bef6ad5702200bc20eac))
* Updated CSS `font-family` to '`helvetica, arial, sans-serif`'. ([#176](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/pull/176), [`1fe10ac`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/1fe10ac52099cecffa997aa72c370e2bdda6c44d))
* Removed redundant CSS rules. ([#176](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/pull/176), [`1fe10ac`](https://github.com/VerifiedJoseph/Save-to-the-Wayback-Machine/commit/1fe10ac52099cecffa997aa72c370e2bdda6c44d))


src/javascript/settings.js
/*jslint node: true */
/*global browser, Intl */

"use strict";

/*
	User Settings
*/
function Settings() {
	var loaded = false,
		self = this,
		items = {},
		defaults = {
			// Page Archived and number format
			numberFormat: '1,000', // Format to display numbers in.

			logNumberArchived: true, // Log number of pages archived.

			// Date and time formats
			dateFormat: 'MMMM d, Y', // Default date format (e.g: May 17, 2019)
			timeFormat: 'h:mm a', // Default date time (e.g: 04:16 PM)
			timeZone: Intl.DateTimeFormat().resolvedOptions().timeZone, // Default timezone from system
			displayFullDate: true, // Display full time and date using user selected formats
			displayTimeSince: false, // Display time past since date (e.g: "1 hour ago")

			// Context Menu
			contextMenu: true, // Enable context menu options.
			contextMenuArchive: { // Enable each context menu options.
				'page': true,
				'link': true,
				'image': true
			},
			contextMenuViewArchived: { // View archived content context menu options
				'page': true,
				'link': true,
				'image': true
			},
			contextMenuNote: true, // Display notification if a page, link or image cannot be archived via the context menu.

			// Notifications
			notePlayAlert: false, // Play an alert sound when a notification is displayed.
			noteAlertSound: 0, // Default Alert sound.

			// Tab & Link Behavior
			openInCurrent: false, // Open archive.org links in the current tab.

			// Debug
			logDebugInfo: false, // Log debug messages in the developer console.

			eolRead: false // End-user has seen and acknowledged the end-of-life message
		};

	/**
	 * Have the settings been loaded?
	 * @returns {boolean}
	 */
	this.isLoaded = function isLoaded() {
		return self.loaded;
	};

	/**
	 * Load user settings from storage
	 * @param {callable} callback
	 */
	this.load = function load(callback) {

		browser.storage.sync.get(defaults, function (data) {
			self.loaded = true;
			self.items = data;

			callback();
		});
	};

	/**
	 * Save updated user settings to storage
	 * @param {object} updatedItems
	 * @param {callback} callback
	 * @callback {boolean} status
	 */
	this.update = function update(updatedItems, callback) {
		var status = true;

		if (typeof updatedItems !== 'undefined') {
			browser.storage.sync.set(updatedItems, function () {
				self.items = updatedItems;

				if (browser.runtime.lastError) {
					status = false;
				}
			});
		}

		callback(status);
	};

	/**
	 * Get setting from the items object 
	 * @param {string} name Name of the setting
	 * @return {mixed}
	 */
	this.get = function get(name) {

		// If item exists in object
		if (self.items.hasOwnProperty(name)) {
			return self.items[name];

		} else { // Item not found
			return null;
		}
	};

	/**
	 * Return all settings (user settings and defaults)
	 * @return {object} all
	 */
	this.getAll = function getAll() {
		var all = {
			options: self.items,
			defaultOptions: defaults
		};

		return all;
	};

}
